[TensorExpr] Add some graph-rewrite passes to prepare models for AOT compilation.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70475
* #70375
* #70374
* #66847
* #70535
* __->__ #66515
* #66514

These passes should not be used generally as they change API of the
model's forward method, but they help experimenting with the model and
ironing out all the kinks before it can be compiled properly. In the
long run ideally we should provide a better way to enable such
experiments.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

Commit:
Add some graph-rewrite passes to prepare BT model for compilation.

Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

[ghstack-poisoned]

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "Add some graph-rewrite passes to prepare BT model for compilation."


Is it a hack?
Yes, we won't be able to use this everywhere, probably we'll have to
require users to adapt their model manually.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "[TensorExpr] Add some graph-rewrite passes to prepare models for AOT compilation."


These passes should not be used generally as they change API of the
model's forward method, but they help experimenting with the model and
ironing out all the kinks before it can be compiled properly. In the
long run ideally we should provide a better way to enable such
experiments.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "[TensorExpr] Add some graph-rewrite passes to prepare models for AOT compilation."


These passes should not be used generally as they change API of the
model's forward method, but they help experimenting with the model and
ironing out all the kinks before it can be compiled properly. In the
long run ideally we should provide a better way to enable such
experiments.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "[TensorExpr] Add some graph-rewrite passes to prepare models for AOT compilation."


These passes should not be used generally as they change API of the
model's forward method, but they help experimenting with the model and
ironing out all the kinks before it can be compiled properly. In the
long run ideally we should provide a better way to enable such
experiments.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "[TensorExpr] Add some graph-rewrite passes to prepare models for AOT compilation."


These passes should not be used generally as they change API of the
model's forward method, but they help experimenting with the model and
ironing out all the kinks before it can be compiled properly. In the
long run ideally we should provide a better way to enable such
experiments.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "[TensorExpr] Add some graph-rewrite passes to prepare models for AOT compilation."


These passes should not be used generally as they change API of the
model's forward method, but they help experimenting with the model and
ironing out all the kinks before it can be compiled properly. In the
long run ideally we should provide a better way to enable such
experiments.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

navahgar commented on torch/csrc/jit/tensorexpr/graph_opt.cpp:
```diff
@@ -402,6 +402,23 @@ void fixupMissingShapeInfo(const std::shared_ptr<Graph>& graph) {
+std::shared_ptr<Graph> replaceListOutputWithTuple(
+    const std::shared_ptr<Graph>& graph) {
+  auto out = graph->outputs()[0];
+  auto out_node = out->node();
```
Can you add a check to ensure that this output node is actually a list here?

navahgar reviewed (approved):
LGTM.

What about tests for these?

Commit:
Update on "[TensorExpr] Add some graph-rewrite passes to prepare models for AOT compilation."


These passes should not be used generally as they change API of the
model's forward method, but they help experimenting with the model and
ironing out all the kinks before it can be compiled properly. In the
long run ideally we should provide a better way to enable such
experiments.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).

Commit:
Update on "[TensorExpr] Add some graph-rewrite passes to prepare models for AOT compilation."


These passes should not be used generally as they change API of the
model's forward method, but they help experimenting with the model and
ironing out all the kinks before it can be compiled properly. In the
long run ideally we should provide a better way to enable such
experiments.

Differential Revision: [D31590862](https://our.internmc.facebook.com/intern/diff/D31590862)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590862).
